### ISSUE TITLE
cmd/compile: Go 1.8 generates poor code for zeroing a small struct

### ISSUE THREAD

**Issue by jacobsa** [golang/go#18370]
I'm comparing Go 1.7.4 and Go 1.8 beta 2 on Darwin:

    go version go1.7.4 darwin/amd64
    go version devel +9cd3c0662a Thu Dec 15 20:06:07 2016 +0000 darwin/amd64

Given this program in `foo.go`:

```Go
package foo

type outHeader struct {
        a uint32
        b int32
        c uint64
}

type OutMessage struct {
        payloadOffset int
        header        outHeader
        payload       [1 << 20]byte
}

func (m *OutMessage) Reset() {
        // Zero the header and payload offset.
        m.payloadOffset = 0
        m.header = outHeader{}
}
```

Go 1.7.4 generates reasonable code for `OutMessage.Reset`, using three 64-bit writes to zero the necessary data:

```
% go tool compile -S foo.go
"".(*OutMessage).Reset t=1 size=29 args=0x8 locals=0x0
        0x0000 00000 (foo.go:15)        TEXT    "".(*OutMessage).Reset(SB), $0-8
        0x0000 00000 (foo.go:15)        NOP
        0x0000 00000 (foo.go:15)        NOP
        0x0000 00000 (foo.go:15)        FUNCDATA        $0, gclocals·87d20ce1b58390b294df80b886db78bf(SB)
        0x0000 00000 (foo.go:15)        FUNCDATA        $1, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
        0x0000 00000 (foo.go:17)        MOVQ    "".m+8(FP), AX
        0x0005 00005 (foo.go:17)        MOVQ    $0, (AX)
        0x000c 00012 (foo.go:18)        MOVQ    $0, 8(AX)
        0x0014 00020 (foo.go:18)        MOVQ    $0, 16(AX)
        0x001c 00028 (foo.go:19)        RET
        0x0000 48 8b 44 24 08 48 c7 00 00 00 00 00 48 c7 40 08  H.D$.H......H.@.
        0x0010 00 00 00 00 48 c7 40 10 00 00 00 00 c3           ....H.@......
[...]
```

However, Go 1.8 beta 2 apparently stores the zero struct in a static variable, that it loads with three separate instructions and then writes back out:

```
% go tool compile -S foo.go
"".(*OutMessage).Reset t=1 size=42 args=0x8 locals=0x0
        0x0000 00000 (foo.go:15)        TEXT    "".(*OutMessage).Reset(SB), $0-8
        0x0000 00000 (foo.go:15)        FUNCDATA        $0, gclocals·a36216b97439c93dafebe03e7f0808b5(SB)
        0x0000 00000 (foo.go:15)        FUNCDATA        $1, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
        0x0000 00000 (foo.go:17)        MOVQ    "".m+8(FP), AX
        0x0005 00005 (foo.go:17)        MOVQ    $0, (AX)
        0x000c 00012 (foo.go:18)        MOVL    "".statictmp_0(SB), CX
        0x0012 00018 (foo.go:18)        MOVL    "".statictmp_0+4(SB), DX
        0x0018 00024 (foo.go:18)        MOVQ    "".statictmp_0+8(SB), BX
        0x001f 00031 (foo.go:18)        MOVL    CX, 8(AX)
        0x0022 00034 (foo.go:18)        MOVL    DX, 12(AX)
        0x0025 00037 (foo.go:18)        MOVQ    BX, 16(AX)
        0x0029 00041 (foo.go:19)        RET
        0x0000 48 8b 44 24 08 48 c7 00 00 00 00 00 8b 0d 00 00  H.D$.H..........
        0x0010 00 00 8b 15 00 00 00 00 48 8b 1d 00 00 00 00 89  ........H.......
        0x0020 48 08 89 50 0c 48 89 58 10 c3                    H..P.H.X..
        rel 14+4 t=15 "".statictmp_0+0
        rel 20+4 t=15 "".statictmp_0+4
        rel 27+4 t=15 "".statictmp_0+8
[...]
```

This shows up for me as poor performance in [real code](https://github.com/jacobsa/fuse/blob/bd496ea0480b848cf655670677b30f2e27702fe4/internal/buffer/out_message.go#L55-L60) when trying out Go 1.8. The effect persists even when the code is linked into a binary.

**Comment by jacobsa** [comment]
I've worked around this for now in my performance-sensitive code using the obvious method:

```Go
func (m *OutMessage) Reset2() {
        m.payloadOffset = 0

        a := (*[2]uint64)(unsafe.Pointer(&m.header))
        a[0] = 0
        a[1] = 0
}
```

which generates

```
"".(*OutMessage).Reset2 t=1 size=35 args=0x8 locals=0x0
        0x0000 00000 (foo.go:23)        TEXT    "".(*OutMessage).Reset2(SB), $0-8
        0x0000 00000 (foo.go:23)        FUNCDATA        $0, gclocals·a36216b97439c93dafebe03e7f0808b5(SB)
        0x0000 00000 (foo.go:23)        FUNCDATA        $1, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
        0x0000 00000 (foo.go:25)        MOVQ    "".m+8(FP), AX
        0x0005 00005 (foo.go:25)        MOVQ    $0, (AX)
        0x000c 00012 (foo.go:27)        LEAQ    8(AX), CX
        0x0010 00016 (foo.go:28)        TESTB   AL, (CX)
        0x0012 00018 (foo.go:28)        MOVQ    $0, 8(AX)
        0x001a 00026 (foo.go:29)        MOVQ    $0, 16(AX)
        0x0022 00034 (foo.go:30)        RET
```

I would be curious to know why that `LEAQ/TESTB` pair is there. (I've seen apparently stray `TESTB` instructions in a few places. Should I file a separate issue for this?)

**Comment by jacobsa** [comment]
(Seems like the `LEAQ/TESTB` is there due to the dereference implicit in `m.header`, because when I type-pun using just `unsafe.Pointer(m)`, it goes away. Not sure why it needs to be there, given the dereference above, but of course I'm in weird territory once I'm type punning.)

**Comment by randall77** [comment]
This looks like a change in the frontend which ends up not triggering the zeroing optimization during translation to SSA.

Here's a simple repro:
```
type X struct {
	a, b, c int
}
func f(x *X) {
	*x = X{}
}
func g(x *X) {
	*x = X{3, 4, 5}
}
```

Go 1.7 does the right thing for `f`, and Go 1.8 does not.
Neither does the right thing for `g`.

For `f`, it looks like the frontend changed to pass a different Node structure to the SSA builder.  In go.1.7, the '*x=X{}' assignment generates:
```
.   AS u(2) l(8) tc(1)
.   .   IND u(2) l(8) tc(1) assigned main.X
.   .   .   NAME-main.x u(1) a(true) g(1) l(7) x(0+0) class(PPARAM) f(1) esc(no) tc(1) used(true) PTR64-*X
.   .   STRUCTLIT l(8) tc(1) main.X
.   .   .   TYPE main.X u(1) a(true) l(3) x(0+0) class(PEXTERN) tc(1) type=main.X main.X main.X
```
In go 1.8 it is:
```
.   AS u(100) l(8) tc(1)
.   .   IND u(100) l(8) tc(1) assigned main.X
.   .   .   NAME-main.x u(1) a(true) g(1) l(7) x(0) class(PPARAM) f(1) esc(no) tc(1) used(true) PTR64-*X
.   .   NAME-main.statictmp_0 u(1) a(true) l(8) x(0) class(PEXTERN) f(1) tc(1) used(true) main.X
```

The SSA translator knows that a `STRUCTLIT` with no args is all zero and optimizes for zeroing.  A `statictmp` could be any value and thus SSA uses a copy from static data.

@griesemer @mdempsky Any idea why the frontend is using a static temp for an all-zero literal?  And any idea why this changed from 1.7 to 1.8?

We could fix that (too late for 1.8?  Probably, unless the fix is really simple).  But the larger problem is `g`.  If it wants to, the SSA backend should really be able to look into the `statictmp` and pull out the static initialization values.  That would fix both `f` and `g`.

**Comment by randall77** [comment]
The `LEAQ/TESTB` code is the nil pointer check.
It looks like we don't have any rule for "if p passes a nil pointer check, so does p+X".
There's even a TODO in cmd/compile/internal/ssa/nilcheck.go:
```
	// TODO: Eliminate more nil checks.
	// We can recursively remove any chain of fixed offset calculations,
	// i.e. struct fields and array elements, even with non-constant
	// indices: x is non-nil iff x.a.b[i].c is.
```
You could work around this for now by doing (untested):
```
func (m *OutMessage) Reset2() {
        m.payloadOffset = 0
        a := (*[3]uint64)(unsafe.Pointer(m))
        a[1] = 0
        a[2] = 0
}
```

**Comment by jacobsa** [comment]
Thanks for looking so quickly!

Regarding `*[3]uint64`: that's what I wound up doing, yeah. Interestingly, though it caused the same code to be generated as Go 1.7.4, my benchmark was still slower than with the `statictmp` version (which in turn is slower than Go 1.7.4). Weird; I can't see a way that eight memory accesses would be better than three.

**Comment by randall77** [comment]
@jacobsa , benchmarking is hard :(

**Comment by jacobsa** [comment]
Yeah, I figure that's it.

**Comment by jacobsa** [comment]
@randall77: I bisected to find that c39918a04991387f14cab1204f54fafab81bc105 is the culprit.

**Comment by bradfitz** [comment]
I'll squawk my typical testing line: if we had had better codegen test coverage, we could stop regressing on things like this. Yes, I know designing the test framework would be the hard part (so we don't need to re-baseline tons of golden tests with every compiler change). The tests would need to describe the general pattern of what we're aim to generate, and not the exact output.

**Comment by jacobsa** [comment]
+1 to @bradfitz; I was going to ask next about whether a regression test could be added.

**Comment by mdempsky** [comment]
It appears to be because of c39918a's change to walk.go:724. (/cc @aclements)

In 1.7.3, we recognized `m.header = outHeader{}` as a zero'ing assignment, and left it as is for SSA to handle.

At tip, we're walk'ing the `outHeader{}` composite literal, which causes us to emit it as a statictmp.

**Comment by aclements** [comment]
> It appears to be because of c39918a's change to walk.go:724. (/cc @aclements)

This was certainly not an intended effect of that change. Since there are no pointers in the structure being zeroed, we should be able to keep doing what we were doing in 1.7 (if there were pointers, it would be a different story).

**Comment by mdempsky** [comment]
CL 34564 is an easy/non-intrusive fix, but it's only a partial solution. It works in this case because outHeader doesn't have any pointer slots.

I think the more complete (but also more intrusive) fix is to just call needwritebarrier in ssa.go. Then we can get rid of OSAWB and applywritebarrier, which will also let us simplify composite literal assignments. I'll prepare a CL for this too, but I think it'll probably want to wait for 1.9.

**Comment by jacobsa** [comment]
Thanks very much for fixing this so quickly, and even adding a test! I'm really impressed how well you all respond to bugs like this.

**Comment by mdempsky** [comment]
@dsymonds FYI in case you missed my comment on CL 34566, the specific case reported here (assignment of zero-value pointer-free composite literals) has been fixed for 1.8.

Assignments of zero-value pointerful composite literals still use statictmp. But unless we have real world examples of this affecting performance critical code, we don't plan to address this for 1.8.

**Comment by dsymonds** [comment]
@mdempsky: I did miss that comment. Thanks! Good to hear.
